core/src/apps/management/sd_protect.py

```diff
@@ -0,0 +1,157 @@
+from trezor.ui.text import Text
+
+from apps.common.confirm import require_confirm
+from apps.common.request_pin import request_pin_ack, request_pin_and_sd_salt
+from apps.common.sd_salt import (
+    SD_SALT_AUTH_KEY_LEN_BYTES,
+    SD_SALT_AUTH_TAG_LEN_BYTES,
+    SD_SALT_LEN_BYTES,
+    commit_sd_salt,
+    remove_sd_salt,
+    set_sd_salt,
+    stage_sd_salt,
+)
+from apps.common.storage import device, is_initialized
+
+if False:
+    from trezor.messages.SdProtect import SdProtect
+
+
+async def sd_protect(ctx: wire.Context, msg: SdProtect) -> Success:
+    if not is_initialized():
+        raise wire.ProcessError("Device is not initialized")
+
+    if msg.operation == SdProtectOperationType.ENABLE:
+        return await sd_protect_enable(ctx, msg)
+    elif msg.operation == SdProtectOperationType.DISABLE:
+        return await sd_protect_disable(ctx, msg)
+    elif msg.operation == SdProtectOperationType.REFRESH:
+        return await sd_protect_refresh(ctx, msg)
+    else:
+        raise wire.ProcessError("Unknown operation")
+
+
+async def sd_protect_enable(ctx: wire.Context, msg: SdProtect) -> Success:
+    salt_auth_key = device.get_sd_salt_auth_key()
+    if salt_auth_key is not None:
+        raise wire.ProcessError("SD card protection already enabled")
+
+    # Confirm that user wants to proceed with the operation.
+    await require_confirm_sd_protect(ctx, msg)
+
+    # Get the current PIN.
+    if config.has_pin():
+        pin = pin_to_int(await request_pin_ack(ctx, "Enter PIN", config.get_pin_rem()))
+    else:
+        pin = pin_to_int("")
+
+    # Check PIN and write salt.
+    salt = random.bytes(SD_SALT_LEN_BYTES)
+    salt_auth_key = random.bytes(SD_SALT_AUTH_KEY_LEN_BYTES)
+    salt_tag = hmac.new(salt_auth_key, salt, sha256).digest()[
+        :SD_SALT_AUTH_TAG_LEN_BYTES
+    ]
+    try:
+        await set_sd_salt(ctx, salt, salt_tag)
+    except Exception:
+        raise wire.ProcessError("Failed to write to SD card")
+
+    if not config.change_pin(pin, pin, None, salt):
+        try:
+            await remove_sd_salt(ctx)
+        except Exception:
+            pass
+        raise wire.PinInvalid("PIN invalid")
+
+    device.set_sd_salt_auth_key(salt_auth_key)
+
+    return Success(message="SD card protection enabled")
+
+
+async def sd_protect_disable(ctx: wire.Context, msg: SdProtect) -> Success:
+    if device.get_sd_salt_auth_key() is None:
+        raise wire.ProcessError("SD card protection not enabled")
+
+    # Confirm that user wants to proceed with the operation.
+    await require_confirm_sd_protect(ctx, msg)
+
+    # Get the current PIN and salt from the SD card.
+    pin, salt = await request_pin_and_sd_salt(ctx, "Enter PIN")
+
+    # Check PIN and remove salt.
+    if not config.change_pin(pin_to_int(pin), pin_to_int(pin), salt, None):
+        raise wire.PinInvalid("PIN invalid")
+
+    device.set_sd_salt_auth_key(None)
+
+    try:
+        await remove_sd_salt(ctx)
+    except Exception:
```

Comment:
This is not very readable. Should it catch `SdProtectCancelled` only?

Reply:
Same as above, the removal is non-critical clean-up operation. Even if anything about the removal of the file fails, the overall operation was a success.